File: Tools/GitHub/ReadMeGenerator.go
package main

import (
	"Documents/Git/OkayegTeaTimeCSharp/Tools/GitHub/Json"
	"fmt"
	"os"
	"strings"

	linq "github.com/ahmetb/go-linq"
)

var (
	header1Text = []string{
		"If your channel has a custom prefix set, commands will have to start with the prefix.",
		"If your channel has no prefix set, commands will have to end with \"eg\", for example: \"pingeg\".",
		"Text in \"[ ]\" is a variable parameter",
	}
	cmdTableHeader = []string{
		"Command",
		"Alias",
		"Description [Parameter | Output]",
	}
	afkCmdTableHeader = []string{
		"Command",
		"Alias",
		"Parameter",
		"Description",
	}
)

const (
	ReadMePath = "../../README.md"
	JsonPath   = "../../OkayegTeaTime/Resources/Commands.json"
	Title      = "OkayegTeaTime"
	Header1    = "Commands"
	Header2    = "AFK-Commands"
)

func main() {
	jsonController := Json.New(JsonPath)
	jsonController.LoadData()
	readMe := GenerateReadMe(jsonController)
	WriteToFile(readMe)
	fmt.Println("Generated README.md")
}

func WriteToFile(input string) {
	err := os.WriteFile(ReadMePath, []byte(input), os.ModePerm)
	if err != nil {
		fmt.Println(err)
	}
}

func GenerateReadMe(jsonController *Json.JsonController) string {
	builder := new(strings.Builder)
	builder.WriteString(fmt.Sprintf("<h1>%s</h1><h2>%s</h2>", Title, Header1))
	linq.From(header1Text).ForEachT(func(s string) {
		builder.WriteString(fmt.Sprintf("%s<br />", s))
	})
	builder.WriteString("<br /><table><tr>")
	linq.From(cmdTableHeader).ForEachT(func(s string) {
		builder.WriteString(fmt.Sprintf("<th>%s</th>", s))
	})
	builder.WriteString("<tr/>")
	linq.From(jsonController.JsonCommands.Commands).SortT(func(c1 Json.Command, c2 Json.Command) bool {
		return c1.CommandName < c2.CommandName
	}).ForEachT(func(c Json.Command) {
		if c.Document {
			builder.WriteString(fmt.Sprintf("<tr><td>%s</td><td><table>", c.CommandName))
			linq.From(c.Alias).ForEachT(func(a string) {
				builder.WriteString(fmt.Sprintf("<tr><td>%s</td></tr>", a))
			})
			builder.WriteString("</table></td><td><table>")
			for i := 0; i < len(c.Parameter); i++ {
				builder.WriteString(fmt.Sprintf("<tr><td>%s</td><td>%s</td></tr>", c.Parameter[i], c.Description[i]))
			}
			builder.WriteString("</table></td></tr>")
		}
	})
	builder.WriteString(fmt.Sprintf("</table><h2>%s</h2><table><tr>", Header2))
	linq.From(afkCmdTableHeader).ForEachT(func(s string) {
		builder.WriteString(fmt.Sprintf("<th>%s</th>", s))
	})
	builder.WriteString("</tr>")
	linq.From(jsonController.JsonCommands.AfkCommands).SortT(func(c1 Json.AfkCommand, c2 Json.AfkCommand) bool {
		return c1.CommandName < c2.CommandName
	}).ForEachT(func(c Json.AfkCommand) {
		builder.WriteString(fmt.Sprintf("<tr><td>%s</td><td><table>", c.CommandName))
		linq.From(c.Alias).ForEachT(func(s string) {
			builder.WriteString(fmt.Sprintf("<tr><td>%s</td></tr>", s))
		})
		builder.WriteString("</table></td>")
		for i := 0; i < len(c.Parameter); i++ {
			builder.WriteString(fmt.Sprintf("<td>%s</td><td>%s</td>", c.Parameter[i], c.Description[i]))
		}
		builder.WriteString("</td></tr>")
	})
	builder.WriteString("</table>")
	return builder.String()
}

File: OkayegTeaTime.Tools/ReadMeGenerator.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using OkayegTeaTime.Models.Json;
using OkayegTeaTime.Twitch.Controller;

namespace OkayegTeaTime.Tools;

public sealed class ReadMeGenerator
{
    private readonly string[] _header1Text =
    {
        "If your channel has a custom prefix set, commands will have to start with the prefix.",
        "If your channel has no prefix set, commands will have to end with \"eg\", for example: \"pingeg\".",
        "Text in \"[ ]\" is a variable parameter"
    };

    private readonly string[] _cmdTableHeader =
    {
        "Command",
        "Alias",
        "Description [Parameter | Output]"
    };

    private readonly string[] _afkCmdTableHeader =
    {
        "Command",
        "Alias",
        "Parameter",
        "Description"
    };

    private const string _title = "OkayegTeaTime";
    private const string _header1 = "Commands";
    private const string _header2 = "Afk-Commands";

    private const string _readMePath = "./README.md";

    public void Generate()
    {
        string readme = CreateReadMe();
        File.WriteAllText(_readMePath, readme);
        Console.WriteLine($"Created README to file: {_readMePath}");
    }

    private string CreateReadMe()
    {
        StringBuilder builder = new();
        CommandController commandController = new();

        builder.Append($"<h1>{_title}</h1><h2>{_header1}</h2>");
        foreach (string headerText in _header1Text)
        {
            builder.Append($"{headerText}<br/>");
        }

        builder.Append("<br/><table><tr>");
        foreach (string cmdHeader in _cmdTableHeader)
        {
            builder.Append($"<th>{cmdHeader}</th>");
        }

        builder.Append("<tr/>");
        Command[] commands = commandController.Commands.Where(c => c.Document).OrderBy(c => c.Name).ToArray();
        foreach (Command cmd in commands)
        {
            builder.Append($"<tr><td>{cmd.Name}</td><td><table>");
            foreach (string alias in cmd.Aliases)
            {
                builder.Append($"<tr><td>{alias}</td></tr>");
            }

            builder.Append("</table></td><td><table>");
            for (int i = 0; i < cmd.Parameters.Length; i++)
            {
                builder.Append($"<tr><td>{cmd.Parameters[i]}</td><td>{cmd.ParameterDescriptions[i]}</td></tr>");
            }

            builder.Append("</table></td></tr>");
        }

        builder.Append($"</table><h2>{_header2}</h2><table><tr>");
        foreach (string cmdHeader in _afkCmdTableHeader)
        {
            builder.Append($"<th>{cmdHeader}</th>");
        }

        builder.Append("</tr>");
        AfkCommand[] afkCommands = commandController.AfkCommands.Where(c => c.Document).OrderBy(c => c.Name).ToArray();
        foreach (AfkCommand cmd in afkCommands)
        {
            builder.Append($"<tr><td>{cmd.Name}</td><td><table>");
            foreach (string alias in cmd.Aliases)
            {
                builder.Append($"<tr><td>{alias}</td></tr>");
            }

            builder.Append("</table></td>");
            for (int i = 0; i < cmd.Parameters.Length; i++)
            {
                builder.Append($"<td>{cmd.Parameters[i]}</td><td>{cmd.ParameterDescriptions[i]}</td>");
            }

            builder.Append("</td></tr>");
        }

        builder.Append("</table>");
        return builder.ToString();
    }
}
